< [1 The Basics of Python](1-TheBasics.ipynb) | [Contents](0-Contents.ipynb) | [3 Control Structures](3-ControlStructures.ipynb) >

# 2 Basic Datatypes
## 2.1 Basic datatypes
There are a few basic datatypes in Python. We already encountered two of them: integers (abbreviated `int`) and decimal numbers (abbreviated `float`). But there are more.

In [1]:
a = 5       # This is an int (whole number)
b = 4.4     # This is a float (with a decimal point)

#### 2.1.1 Logical values
A third datatype is boolean (or logical, abbreviated `bool`). When you **compare** numbers, strings, etc. the result is either `True` (1) or `False` (0). Check the following examples:

In [2]:
bool0 = True
bool1 = 5 > -5
bool2 = 6**2 < 2**5

In [3]:
print(bool1)
print(bool2)

True
False


#### 2.1.2 Strings
Another datatype is string (abbreviated `str`). We already met strings in the previous chapter as the variable `s = "The time is 13:24."`

A string is a combination of letters and numbers and other characters (punctuation marks, special characters, etc.). Whenever you place letters and numbers in between (single or double) quotes you get a string.

Have a look at the following examples (execute the code cells):

In [4]:
"This is a string with 7 words!"

'This is a string with 7 words!'

In [5]:
"9876"

'9876'

In [6]:
"3.14159"

'3.14159'

| type | Meaning |
| :---: | :---: |
| `int` | Integer (whole number) |
| `float` | Float (decimal number) |
| `bool` | Boolean (True or False) |
| `str` | String (Piece of text) |


Note that even though we typed only numbers when entering 9876 we still have a string because of the double quotes.

To check the datatype we can use the built in `type()` function:

In [7]:
type("9876")

str

Note the difference between this and the next example.

In [8]:
type(9876)

int

In [9]:
type(3.14159)

float

In [10]:
type(False)

bool

Type conversion is possible with the built in functions `int()`, `float()`, `str()` and `bool()`. For example:

In [ ]:
a = "5.0"
b = int(a)
type(b)
print(b)

#### Exercise on type conversion

##### Exercise 1
Fill in the dots in the following code fragment with the necessary functions for type conversion (also use the `round` function) so that the displayed output appears correctly on the screen.

**Output**

```
Mean temperature June: 16.57 degrees.
After rounding, this is 16.6 degrees.
Mean temperature July: 18.66 degrees.
After rounding, this is 18.7 degrees.
```


In [ ]:
temp_juni = 16.57
temp_juli = "18.66"
b1 = "Mean temperature June: " + ....... + " degrees"
b2 = "After rounding, this is " + ....... + " degrees"
b3 = "Mean temperature July: " + ....... + " degrees"
b4 = "After rounding, this is " + ....... + " degrees"
print (b1 , b2 , b3 , b4 , sep = "\n")


##### Exercise 2
Think about what the output will be after executing the following commands:

In [12]:
print(int(True))
print(int(False))
print(bool(0))
print(bool(1))
print(bool(25))

1
0
False
True
True


#### 2.2.3 String concatenation and duplication
With the `+` and `*` operators strings can be **concatened** and **duplicated**.

In [14]:
# giving variables text, and adding text.
word1 = "Good"
word2 = "morning"
word3 = "to you too!"
# string concatenation with the print() function
print(word1, word2)
print(word1, word3)

Good morning
Good to you too!


In [ ]:
print(word1, word2, sep = "+++", end = "---")
print(word1, word3)

In [ ]:
# string concatenation with the + operator
sentence1 = word1 + word2 + word3
print(sentence1)
sentence2 = word1 + " " + word2 + " " + word3
print(sentence2)

In [ ]:
# string duplication with the * operator
sentence3 = 3*word1
print(sentence3)
print("-"*12)

We observe that:
* the `print()` adds one space in between strings. This can be changed with the `sep` argument of the `print()` function.
* At the end of the line, `print()` adds a newline character. This can be changed with the `end` argument of the `print()` function.
* with the `+` operator strings are concatenated without extra spaces in between
* string duplication with the `*` operator saves typing time

The `len()` function gives the **number of characters** in a string.:

In [16]:
text = "abc def, ghij!"
len(text)

14

Note that the spaces, the comma and the exclamation mark also take places: the 14 we obtained is the sum of
* 10 letters
* 2 spaces
* 1 exclamation mark
* 1 comma

#### 2.2.4 String manipulation
Often we need to manipulate strings:
* search for a character or a substring,
* count the numberof occurances,
* replace one or more characters by others,
* split the string,
* etc.

Python has an extensive list of string methods. Some useful methods are:

| Method | Description |
| :---: | --- |
| `find()` | Searches the string for a specified value and returns the **first** position of where it was found |
| `count()` | Returns the number of times a specified value occurs in a string |
| `replace()` | Returns a string where a specified value is replaced with a specified value |
| `split()` | Splits the string at the specified separator, and returns a list |
| `lower()` | Converts a string into **lower** case |
| `upper()` | Converts a string into **upper** case |

The following examples illustrate the use of these string methods:

In [15]:
word1 = "Good"
word2 = "morning"
word3 = "to you too!"

print(word1.count("o"))

2


In [17]:
print(word2.upper())

MORNING


In [18]:
print(word2)

morning


We notice that the `.upper()` method does not change the original string `s` but returns a new string. The same applies to all string methods: they do not change the original string but return a new string.

In [19]:
print(word3.split(" "))

['to', 'you', 'too!']


In [20]:
print(word3.replace("too!", "my friend."))

to you my friend.


#### 2.2.5 Accessing Elements in a string
In many applications we have to address a single character or make selections from text. Strings are similar to `lists` which we will introduce later. Similar operations (called *list slicing*) apply to strings.

#### Indexing
Each character in a string has an index. Indexing starts at 0:
* the first (position 1) character has index 0,
* the second (position 2) character has index 1, etc.

Access to a character is granted by the use of the **index operator** `[]`:

```Python
    s[i] # with i ranging from 0 to len(s)-1
```
means: select the character with index `i` (position `i+1`) from the string `s`. The index `i` ranges from `0` to `len(s)-1`.

Examples:

In [21]:
s = "abcdefghij"
print(s[0])    # FIRST character

a


In [22]:
print(s[2])    # THIRD character

c


Note that index `2` refers to the character at position `3`.

Indexing with **negative indices** is also possible. The syntax is (`s` is a string):

```Python
    s[-i] # with i ranging from 1 to len(s)
```

In [23]:
print(s[-2])

i


We have selected the second character from the right. Note that here we start counting with -1 for the first character on the right.

The last character of a string can be indexed with a negative index using `-1`, the first character has the negative index `-len(text)`:

In [24]:
s = "abcdefghij"
print(s[-1])         # LAST character, equivalent with s[len(s)-1]

j


In [25]:
print(s[-len(s)]) # FIRST character, equivalent with s[0]

a


#### Slicing
Slicing is the selection of multiple characters. Again, the index operator `[]` is used:

```Python
    s[i:j] # with i < j
```

means: select the characters with indices `i`, `i+1`, up to `j-1`. The character with index `j` is **not included** in the selection.

The **general** syntax is:

```python
    s[start:stop:step]
```
where
* `start` is the index of the first element (default is 0),
* `stop` is the index of the last element (**not included**, default is len(list_name)) and
* `step` is the number of elements to skip (default is 1).

Example:

In [26]:
print(s[0:4])   # select characters with indices 0 up to 3

abcd


The following table summarizes some indexing and slicing possibilities (the variable `s` represents a string):

| Slice | Meaning |
| :---: | --- |
| `s[i]` | character at index `i` (counting starts at 0) |
| `s[i:j]` | characters `i` up to `j-1` (character `j` **not** included) |
| `s[i:]` | characters from `i` up to end of string |
| `s[:j]` | characters from start up to `j-1` (character  `j` **not** included) |
| `s[i:j:k]` | characters from `i` up to `j-1` in steps of `k` |
| `s[:]` | all characters |
| `s[::-1]` | all characters in **reversed** order|

Important note: with ranges, the **stop value is not included in the selection**.

Some examples:

In [27]:
s = "abcdefghij"
print(s[4:8])

efgh


Here we see that `[4:8]` selects characters with indices 4, 5, 6, and 7 which is `efgh`.

In [28]:
print(s[:4])

abcd


In [29]:
print(s[4:])

efghij


With `[4:]` we start with index 4 until the end of the string, which results in `efghij`.

In [30]:
print(s[1:7:3])

be


In the last example only the characters with indices 1 and 4 are selected. The character with index 7 is not included in the selection.

#### 2.1.3 Summary
Let us summarize the datatypes we already know in a table:

| Abbreviation | Datatype | Example(s) |
| :---: | --- | --- |
| `int` | integer | 5, -73 |
| `float` | decimal number | 3.14, -29.469, 4.0 |
| `str` | string | "word", "12.3" |
| `bool` | boolean | `True` (1) and `False` (0) |

### Exercise 1

Execute the code cell below to define and print the variable `sentence1`:

In [31]:
word1 = "Good"
word2 = "morning"
word3 = "to you too!"
sentence1 = word1 + " " + word2 + " " + word3
print(sentence1)


Good morning to you too!


Replace the $\ldots$ in the cell below to select the **second word** of the sentence using the correct **range**.

In [35]:
secondWord = sentence1[5:12]
print(secondWord)

morning


### Exercise 2
Suppose you have a string `name` containing the name of a Python script. Use slicing/indexing to select the file name without the extension `.py`. Replace the dots between the brackets:

In [42]:
name = "intro_to_Python.py"
filename = name[:-3]
print(filename)

intro_to_Python


Now suppose that the string `name` contains a file name with an extension that can be of **variable length**, e.g. `my_script.py`, `data.csv`, `presentation.pptx`, etc. Use slicing/indexing and other string methods to select the file name without the extension.

In [43]:
name = "intro_to_Python.ipynb"
filename = name[0:len(name)-6]
print(filename)


intro_to_Python


### Exercise 3

Consider the following string (typically part of a fasta file – the complete fasta format is more extensive):

>MCHU - Calmodulin - Human, rabbit, bovine, rat, chicken
MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTID
FPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREADIDGDGQVNYEEFVQMMTAK*

Execute the code cell below to define and print the variable `fasta_str`:


In [44]:
fasta_str = ">MCHU - Calmodulin - Human, rabbit, bovine, rat, chicken\nMADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTID\nFPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREADIDGDGQVNYEEFVQMMTAK*"
print(fasta_str)


>MCHU - Calmodulin - Human, rabbit, bovine, rat, chicken
MADQLTEEQIAEFKEAFSLFDKDGDGTITTKELGTVMRSLGQNPTEAELQDMINEVDADGNGTID
FPEFLTMMARKMKDTDSEEEIREAFRVFDKDGNGYISAAELRHVMTNLGEKLTDEEVDEMIREADIDGDGQVNYEEFVQMMTAK*


Note that, even though the string is defined as a single line, it is displayed as multiple lines. Try to explain this behaviour by looking at the string: do you notice any **special characters** in the string?

In [45]:
#The '\n' in python causes the print statement to create a new line in the output

(a) The **first four** characters after the `>` encode the protein. Extract this code from the string.

In [46]:
code = fasta_str[1:5]
print(code)

MCHU


(b) Count the number of times the combination `AE` occurs

In [47]:
print(fasta_str.count("AE"))


3


(c) The part between the the first `-` and the second `-` describes the protein. Use string methods and slicing/indexing to extract this description. Note the **space before and after** the `-` characters in the protein string.

In [49]:
print(fasta_str[7:19])


 Calmodulin 


## 2.5 String formatting with f-strings

With the use of `f` strings variables (containing integers, floats, strings, ...) can be put together in a formatted string. Formatting can be
* the number op characters (places) to be used
* desired precision (number of decimals)
* alignment specification: left, right, center

#### Basic formatting

To use `f` strings:
* place an `f` or an `F` before the string to be formatted
* type accolades `{}` (_place holders_) around the variable to be formatted
* let the variable be followed by a colon `:`
* specify the datatype of the variable:

| Character | Datatype |
| :---: | --- |
| `s` | string |
| `d` | integer |
| `f` | float |

* print the `f` string with the `print()` function

The syntax to format a variable is

```Python
    {variable:xtype}
```

with:
* `variable`: the name of the variable to be formatted
* `x`: the number of characters (places) to be used
* `type`: the datatype of the variable to be formatted (`s`, `d` or `f`)

#### Specifying the number of characters

To specify the number of characters to be used, simply `x` by this number:

In [50]:
char = "a"
n = 23
print(f"The character {char:5s} occurs {n:10d} times")

The character a     occurs         23 times


Observe that:
* since **strings** are aligned **left**, the character `a` is **followed** by 4 additional spaces
* since **numbers** are aligned **right**, the number `23` is **preceded** by 8 additional spaces

#### Precision

In case of decimal numbers (floats), one can specify the precision. This can be realized by using the following syntax:

```Python
    {variable:x.yf}
```

with:
* `variable`: the name of the variable to be formatted
* `x`: the number of characters (places) to be used
* `y`: the number of decimal places to be used

Execute the following code and try to interpret the result:

In [51]:
bmi = 21.718
name = "John"
print(f"{name:10s} has a BMI of {bmi:7.1f}.")

John       has a BMI of    21.7.


The formatting `{name:10s}` means: use 10 characters to represent the value of `name`. Since 10 places were provided and strings are aligned to the left, and `John` takes 4 places, 6 additional spaces were added to the **right** of `John`. The `s` stands for string.

The formatting `{bmi:7.1f}` means: use 7 characters to represent the value of `bmi` and round to 1 decimal place. The `f` stands for float.

Since 7 places were provided, and numbers are aligned to the right, and `21.7` takes 4 places (the dot takes one too), 3 additional spaces were added to the **left** of `21.7`.

#### Alignment
The alignment can be specified by adding `<`, `>` or `^` just after the colon:

| Character | Alignment |
| :---: | --- |
| `<` | left |
| `>` | right |
| `^` | centered |

< [1 The Basics of Python](1-TheBasics.ipynb) | [Contents](0-Contents.ipynb) | [3 Control Structures](3-ControlStructures.ipynb) >